mm2src/coins/qrc20.rs

```diff
@@ -0,0 +1,1004 @@
+            log!("sat_balance " [sat_balance] " min_amount " [min_amount]);
+            if sat_balance < min_amount.into() {
+                return ERR!(
+                    "Base coin balance {} is too low to cover gas fee, required {}",
+                    qtum_balance,
+                    big_decimal_from_sat(min_amount as i64, decimals),
+                );
+            }
+            Ok(())
+        }))
+    }
+
+    fn wallet_only(&self) -> bool { false }
+
+    fn withdraw(&self, req: WithdrawRequest) -> Box<dyn Future<Item = TransactionDetails, Error = String> + Send> {
+        Box::new(qrc20_withdraw(self.clone(), req).boxed().compat())
+    }
+
+    fn decimals(&self) -> u8 { utxo_common::decimals(&self.utxo) }
+
+    fn convert_to_address(&self, from: &str, to_address_format: Json) -> Result<String, String> {
+        let to_address_format: Qrc20AddressFormat =
+            json::from_value(to_address_format).map_err(|e| ERRL!("Error on parse QRC20 address format {:?}", e))?;
+        let from_address = try_s!(self.try_address_from_str(from));
+        match to_address_format {
+            Qrc20AddressFormat::Wallet => Ok(from_address.to_string()),
+            Qrc20AddressFormat::Contract => Ok(display_contract_address(from_address)),
+        }
+    }
+
+    fn validate_address(&self, address: &str) -> ValidateAddressResult { utxo_common::validate_address(self, address) }
+
+    fn process_history_loop(&self, ctx: MmArc) { self.history_loop(ctx) }
+
+    fn history_sync_status(&self) -> HistorySyncState { utxo_common::history_sync_status(&self.utxo) }
+
+    fn get_trade_fee(&self) -> Box<dyn Future<Item = TradeFee, Error = String> + Send> {
+        utxo_common::get_trade_fee(self.clone())
+    }
+
+    fn required_confirmations(&self) -> u64 { utxo_common::required_confirmations(&self.utxo) }
+
+    fn requires_notarization(&self) -> bool { utxo_common::requires_notarization(&self.utxo) }
+
+    fn set_required_confirmations(&self, confirmations: u64) {
+        utxo_common::set_required_confirmations(&self.utxo, confirmations)
+    }
+
+    fn set_requires_notarization(&self, requires_nota: bool) {
+        utxo_common::set_requires_notarization(&self.utxo, requires_nota)
+    }
+
+    fn my_unspendable_balance(&self) -> Box<dyn Future<Item = BigDecimal, Error = String> + Send> {
+        // QRC20 cannot have unspendable balance
+        Box::new(futures01::future::ok(0.into()))
+    }
+}
+
+pub fn qrc20_swap_id(time_lock: u32, secret_hash: &[u8]) -> Vec<u8> {
+    let mut input = vec![];
+    input.extend_from_slice(&time_lock.to_le_bytes());
+    input.extend_from_slice(secret_hash);
+    sha256(&input).to_vec()
+}
+
+pub fn qrc20_addr_from_str(address: &str) -> Result<H160, String> {
+    // use deserialization instead of eth::contract_addr_from_str(),
+    // because that function fails on some of the QRC20 contract addresses
+    Ok(try_s!(json::from_str(&format!("\"{}\"", address))))
```

Comment:
Could you please clarify on what addresses `eth::contract_addr_from_str()` failed?

Reply:
It fails on is_valid_checksum_addr()
For example, our QRC20 testnet token address `0xd362e096e873eb7907e205fadc6175c6fec7bc44` failed with
`"eth:2497] Invalid address checksum"`

Reply:
There is the eth::addr_from_str that doesn't check the checksum. I replace the serde deserialization with this method, thanks!